### PR TITLE
bt-157: add logic to sign in flow

### LINKED ISSUE #157: Add logic to sign in flow
We need to add logic for sign in flow: validation schema, request to the backend (can be mocked if now merged yet) and saving user data to the store. User token should be stored in phone's local storage. After user is signed in we need to check his/her role (Talent, Employer) and redirect to needed navigation (it should be refactored to support 2 roles)

### PR DESCRIPTION
**Pull request by mskmee**

### CODE REVIEW

File: mobile/src/bundles/auth/components/sign-in-form/sign-in-form.tsx [hunk 1]
```diff
@@ -11,19 +11,23 @@ import {
 import { AuthScreenName, TextCategory } from '~/bundles/common/enums/enums';
 import { useAppForm, useCallback } from '~/bundles/common/hooks/hooks';
 import { globalStyles } from '~/bundles/common/styles/global-styles';
+import {
+    type UserSignInRequestDto,
+    userSignInValidationSchema,
+} from '~/bundles/users/users';
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/components/sign-in-form/sign-in-form.tsx]: let's get it from '~/bundles/auth/...

File: mobile/src/bundles/auth/screens/auth.tsx [hunk 2]
```diff
@@ -21,26 +24,27 @@ const Auth: React.FC = () => {
     const { dataStatus } = useAppSelector(({ auth }) => ({
         dataStatus: auth.dataStatus,
     }));
-
     const isSignUpScreen = name === AuthScreenName.SIGN_UP;
-
+    const isLoading = dataStatus === DataStatus.PENDING;
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/screens/auth.tsx]:
```suggestion
    const isPendingAuth = dataStatus === DataStatus.PENDING;
```

File: mobile/src/bundles/auth/store/actions.ts [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
 
+import { ErrorMessages } from '~/bundles/common/enums/enums';
 import { type AsyncThunkConfig } from '~/bundles/common/types/types';
 import {
+    type UserSignInRequestDto,
+    type UserSignInResponseDto,
     type UserSignUpRequestDto,
     type UserSignUpResponseDto,
 } from '~/bundles/users/users';
+import { StorageKey } from '~/framework/storage/enums/storage-key.enum';
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/store/actions.ts]:
```suggestion
import { StorageKey } from '~/framework/storage/enums/enums';
```

File: mobile/src/bundles/auth/store/slice.ts [hunk 4]
```diff
@@ -1,32 +1,61 @@
...
 
-import { signUp } from './actions';
+import { signIn, signUp } from './actions';
+
+type UserData = {
+    email: string | null;
+    id: string | null;
+    role: ValueOf<typeof UserRole> | null;
+    isProfileComplete: boolean;
+};
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/store/slice.ts:14]: Add here todo comment to replace with shared UserData type

File: mobile/src/bundles/auth/store/slice.ts [hunk 5]
```diff
@@ -1,32 +1,61 @@
...
     reducers: {},
     extraReducers(builder) {
-        builder.addCase(signUp.pending, (state) => {
+        builder.addMatcher(
+            isAnyOf(signUp.fulfilled, signIn.fulfilled),
+            (state, { payload }) => {
+                state.dataStatus = DataStatus.FULFILLED;
+                state.userData.email = payload.email;
+                state.userData.id = payload.id;
+                state.userData.role = payload.role;
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/store/slice.ts]:
```suggestion
                state.userData = {email, id, role}
```

File: mobile/src/bundles/auth/store/slice.ts [hunk 6]
```diff
@@ -1,32 +1,61 @@
...
 const { reducer, actions, name } = createSlice({
     initialState,
     name: 'auth',
     reducers: {},
     extraReducers(builder) {
-        builder.addCase(signUp.pending, (state) => {
+        builder.addMatcher(
+            isAnyOf(signUp.fulfilled, signIn.fulfilled),
+            (state, { payload }) => {
+                state.dataStatus = DataStatus.FULFILLED;
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/store/slice.ts:38]:
```suggestion
                const {email, id, role} = payload
                state.dataStatus = DataStatus.FULFILLED;
```

File: mobile/src/framework/api/http-api-base.ts [hunk 7]
```diff
@@ -53,7 +53,6 @@ class HttpApiBase implements HttpApi {
             headers,
             payload,
         });
-
```
Comment (katerynakostikova) [inline mobile/src/framework/api/http-api-base.ts:56]: let's leave line here)

File: mobile/src/navigations/root/root.tsx [hunk 8]
```diff
@@ -51,7 +48,10 @@ const Root: React.FC<Properties> = ({
                     component={AuthNavigator}
                 />
             ))}
+
         </RootStack.Navigator>
+    <Toast />
```
Comment (katerynakostikova) [inline mobile/src/navigations/root/root.tsx]: Why do we have toast here?

File: shared/src/bundles/users/validation-schemas/constants.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+const constants = {
+    MIN_PASSWORD_LENGTH: 8,
+    MIN_EMAIL_LENGTH: 5,
+    MAX_LOGIN_INPUT_LENGTH: 50,
+    PASSWORD_REGEXP:
+        /^(?=.*[A-Z])(?=.*[a-z])(?=.*\d)(?=.*[!#$%&()*+,./:;<=>?@^_{}-]).+$/,
+};
```
Comment (katerynakostikova) [inline shared/src/bundles/users/validation-schemas/constants.ts]:
```suggestion
const AUTH_CONSTANTS = {
    MIN_PASSWORD_LENGTH: 8,
    MIN_EMAIL_LENGTH: 5,
    MAX_LOGIN_INPUT_LENGTH: 50,
    PASSWORD_REGEXP:
        /^(?=.*[A-Z])(?=.*[a-z])(?=.*\d)(?=.*[!#$%&()*+,./:;<=>?@^_{}-]).+$/,
};
```

File: mobile/src/navigations/root/root.tsx (hunk 8 above)
Comment (mskmee) [reply mobile/src/navigations/root/root.tsx]: dont know why I didnt add him to the index ^)

File: mobile/src/bundles/auth/auth-api.ts [hunk 10]
```diff
@@ -35,6 +37,22 @@ class AuthApi extends HttpApiBase {
...
+            this.getFullEndpoint(AuthApiPath.SIGN_IN, {}),
+            {
+                method: 'POST',
+                contentType: ContentType.JSON,
+                payload: JSON.stringify(payload),
+                hasAuth: false,
+            },
+        );
+
+        return await response.json<UserSignUpResponseDto>();
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/auth-api.ts]:
```suggestion
        return await response.json< UserSignInResponseDto >();
```

File: mobile/src/bundles/auth/store/actions.ts [hunk 11]
```diff
@@ -14,8 +18,27 @@ const signUp = createAsyncThunk<
...
+        const response = await authApi.signIn(signInPayload);
+        await storage.set(StorageKey.TOKEN, response.token);
+        return response;
+    } catch (error) {
+        if (error instanceof Error) {
+            notifications.showError(error.message);
+            throw error;
+        }
+        notifications.showError(ErrorMessages.UNKNOWN_ERROR);
+        throw error;
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/store/actions.ts:40]:
```suggestion
```

File: mobile/src/bundles/auth/store/slice.ts [hunk 12]
```diff
@@ -1,32 +1,61 @@
...
+    id: string | null;
+    role: ValueOf<typeof UserRole> | null;
+    isProfileComplete: boolean;
+};
+// TODO: Replace user data with shared UserData Type
 
 type State = {
     dataStatus: ValueOf<typeof DataStatus>;
+    isSignedIn: boolean;
+    userData: UserData;
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/store/slice.ts]:
```suggestion
    userData: UserData | null;
```

File: mobile/src/bundles/auth/store/slice.ts [hunk 13]
```diff
@@ -1,32 +1,61 @@
...
 
 const initialState: State = {
     dataStatus: DataStatus.IDLE,
+    isSignedIn: false,
+    userData: {
+        email: null,
+        id: null,
+        role: null,
+        isProfileComplete: false,
+    },
```
Comment (katerynakostikova) [inline mobile/src/bundles/auth/store/slice.ts]:
```suggestion
    userData: null
```

File: mobile/src/navigations/root/root.tsx [hunk 14]
```diff
@@ -19,25 +21,18 @@ const screenOptions: NativeStackNavigationOptions = {
...
-    isSignedIn = false,
-    isProfileComplete = true,
-    role = 'talent',
-}) => {
+const Root: React.FC = () => {
+    const {
+        isSignedIn,
+        userData: { isProfileComplete, role },
+    } = useAppSelector(({ auth }) => auth);
     // prettier-ignore
```
Comment (katerynakostikova) [inline mobile/src/navigations/root/root.tsx]:
```suggestion
    const {
        isSignedIn,
        userData,
    } = useAppSelector(({ auth }) => auth);
    
    const { isProfileComplete, role} = userData ?? {}
    // prettier-ignore
```

File: mobile/src/bundles/auth/store/actions.ts (hunk 11 above)
Comment (mskmee) [reply mobile/src/bundles/auth/store/actions.ts:40]: Cant do it, then action will return undefined

File: mobile/src/bundles/auth/auth-api.ts (hunk 10 above)
Comment (mskmee) [reply mobile/src/bundles/auth/auth-api.ts]: It's project linter, can disable it only with disable it

File: mobile/src/bundles/auth/store/actions.ts (hunk 11 above)
Comment (katerynakostikova) [reply mobile/src/bundles/auth/store/actions.ts:40]: Ok, sorry, my mistake

File: mobile/src/bundles/auth/auth-api.ts (hunk 10 above)
Comment (katerynakostikova) [reply mobile/src/bundles/auth/auth-api.ts]: ?

Comment (katerynakostikova) [reply mobile/src/bundles/auth/auth-api.ts]: I mean you have UserSignUpResponseDto but should be UserSignInResponseDto

Comment (mskmee) [reply mobile/src/bundles/auth/auth-api.ts]: yeap, my bad (

File: mobile/src/navigations/root/root.tsx [hunk 15]
```diff
@@ -5,7 +5,9 @@ import {
 import React from 'react';
 
 import { RootScreenName } from '~/bundles/common/enums/enums';
+import { useAppSelector } from '~/bundles/common/hooks/use-app-selector/use-app-selector.hook';
```
Comment (katerynakostikova) [inline mobile/src/navigations/root/root.tsx]:
```suggestion
import { useAppSelector } from '~/bundles/common/hooks/hooks';
```

**Review by katerynakostikova — APPROVED** [review]

**Review by illia-dobro — APPROVED** [review]

**Pull request merged**